Source/Scene/Model/MetadataPipelineStage.js

```diff
@@ -62,60 +83,225 @@ MetadataPipelineStage.process = function (
   frameState
 ) {
   const { shaderBuilder, model } = renderResources;
+  const { structuralMetadata = {}, content } = model;
+  const statistics = content?.tileset?.metadataExtension?.statistics;
 
-  const structuralMetadata = defaultValue(model.structuralMetadata, {});
-  const { propertyAttributes, propertyTextures } = structuralMetadata;
-
-  // Find which metadata types are used
-  const metadataTypes = new Set();
-  if (defined(propertyAttributes)) {
-    getPropertyAttributeTypes(metadataTypes, primitive, propertyAttributes);
-  }
-  if (defined(propertyTextures)) {
-    getPropertyTextureTypes(metadataTypes, propertyTextures);
-  }
+  const propertyAttributesInfo = getPropertyAttributesInfo(
+    structuralMetadata.propertyAttributes,
+    primitive,
+    statistics
+  );
+  const propertyTexturesInfo = getPropertyTexturesInfo(
+    structuralMetadata.propertyTextures,
+    statistics
+  );
 
-  // Declare only the needed <type>MetadataClass structs
-  declareMetadataTypeStructs(shaderBuilder, metadataTypes);
+  // Declare <type>MetadataClass and <type>MetadataStatistics structs as needed
+  const allPropertyInfos = propertyAttributesInfo.concat(propertyTexturesInfo);
+  declareMetadataTypeStructs(shaderBuilder, allPropertyInfos);
 
-  // Always declare the Metadata and MetadataClass structs, and the
-  // initializeMetadata() function, even if not used
+  // Always declare the Metadata, MetadataClass, and MetadataStatistics structs
+  // and the initializeMetadata() function, even if not used
   declareStructsAndFunctions(shaderBuilder);
   shaderBuilder.addVertexLines([MetadataStageVS]);
   shaderBuilder.addFragmentLines([MetadataStageFS]);
 
-  if (defined(propertyAttributes)) {
-    processPropertyAttributes(renderResources, primitive, propertyAttributes);
+  for (let i = 0; i < propertyAttributesInfo.length; i++) {
+    const info = propertyAttributesInfo[i];
+    processPropertyAttributeProperty(renderResources, info);
   }
-  if (defined(propertyTextures)) {
-    processPropertyTextures(renderResources, propertyTextures);
+  for (let i = 0; i < propertyTexturesInfo.length; i++) {
+    const info = propertyTexturesInfo[i];
+    processPropertyTextureProperty(renderResources, info);
   }
 };
 
+/**
+ * Collect info about all properties of all propertyAttributes, and
+ * return as a flattened Array
+ * @param {PropertyAttribute[]} propertyAttributes
+ * @param {ModelComponents.Primitive} primitive
+ * @param {Object} statistics
+ * @returns {Array<{Object}>}
```

Comment:
Does `{}` around a type mean something in JSDoc?

Unless that had a specific meaning, I think this can be simplified to this:

```suggestion
 * @returns {Object[]}
```

Reply:
I see this in a few other places in this file, make sure to update those, too.